config/initializers/secure_headers.rb

```diff
@@ -33,3 +33,12 @@
   #   ]
   # }
 end
+
+SecureHeaders::Configuration.override(:saml) do |config|
+  providers = YAML.load_file("#{Rails.root}/config/service_providers.yml")
+  providers = providers.fetch(Rails.env, {})
+  providers.symbolize_keys!
+  whitelisted_urls = providers[:valid_hosts].values.map { |k, _| k['acs_url'] }
+
+  whitelisted_urls.each { |url| config.csp[:form_action] << url }
+end
```

Comment:
The current requesting SP cannot be determined at initialization time. The benefit of this override is that it is only calculated once and reused for each request. Setting the headers dynamically for each request would negatively impact performance, without providing any substantial benefit.

Using the domain name only makes sense. I will make that change.